### ISSUE TITLE
handle filter negation via filters.js

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#1420]
No filters should implement their own negation handlers. Instead, a helper function in **$:/core/modules/filters.js** should evaluate a boolean expression and depending on the negation prefix **!** add to the result-set or not. Every filter should thus be automatically negation-enabled.

**Comment by Jermolene** [comment]
I agree that a helper function for processing filter results is a good idea, but I don't think that all filter implementations should use it. The reason is that in some situations one can process the operator prefix outside of the loop through the source tiddlers. So we shouldn't pay the cost of the checks for all filters.

**Comment by tobibeer** [comment]
Of course, only those filters where that is needed. The benefits simply are...

* code simplification
* making sure a negation indeed is a negation

Something like **removeprefix** possibly won't need a negation.

**Comment by tobibeer** [comment]
Like the proposed `$tw.utils.intersect()` (see #1385), there should perhaps equally be a `$tw.utils.negate(set,subset)` returning the negation of the `subset` with respect to a `set`.

I'm not exactly sure why we wouldn't simply perform this for the result of every filter, whether we anticipate it as useful or not.

**Comment by Jermolene** [comment]
> I'm not exactly sure why we wouldn't simply perform this for the result of every filter, whether we anticipate it as useful or not.

As I said on another commit, the filter operations need to be as fast as possible. Some operations can handle negation much more efficiently than a generic helper like this.

**Comment by pmario** [comment]
@tobibeer ... imo there is consensus, that "speed matters" ... so can we close the issue?

**Comment by tobibeer** [comment]
@pmario: To me, the matter at hand is reliability and logical consistency. From my perspective, those overrule any speed concerns.

**Comment by tobibeer** [comment]
This appears to be a suggested solution for #1454.
